### DISCUSSION TITLE
Non-mainstream rtos mg_connect is used

### DISCUSSION THREAD

**Discussion by evercf** [cesanta/mongoose#2180] [Ask questions, get answers] [answered]
1. static void TCPIP_test(void * argv)
2. {
3.     struct mg_mgr mgr;
4.     //struct mg_connection *c;
5.     char ip4_addr_str[16] = {0};
6.     struct sockaddr_in	ip4_local_addr = {0};
7.     quec_uart_test_tcp();
8.     //datacall_satrt();
9.     //ql_get_data_call_info(1, 0, &info);
10.     strcpy(tcp_data,"========== mg_connect start\r\n");
11.     ql_uart_write(QL_MAIN_UART_PORT, tcp_data, strlen(tcp_data));
12.     mg_mgr_init(&mgr); // Init manager
13.    
14.     if (c_res.c == NULL) {
15.         // connect
16.         c_res.i=0;
17.         c_res.c = mg_connect(&mgr, s_conn, cfn, &c_res);
18.         if (c_res.c == NULL)
19.         {    
20.             strcpy(tcp_data,"CLIENT cant' open a connection\r\n");
21.             ql_uart_write(QL_MAIN_UART_PORT, tcp_data, strlen(tcp_data));
22.         }else
23.         {
24.             strcpy(tcp_data,"CLIENT is connecting\r\n");
25.             ql_uart_write(QL_MAIN_UART_PORT, tcp_data, strlen(tcp_data));
26.         }
27.     }
28.     //mg_timer_add(&mgr, 15000, MG_TIMER_REPEAT | MG_TIMER_RUN_NOW, timer_fn,&mgr);  // Init timer for demo purposes, 15s
29.     
30.     while(true)
31.     {
32.         mg_mgr_poll(&mgr, 100);                 // Event loop
33.     }
34.     mg_mgr_free(&mgr);                                // Cleanup
35. 
36. }
37. 
38. application_init(TCPIP_test, "TCPIP_test", 4, 4);

> The results are as follows：
> ========== mg_connect start
> CLIENT has been initialized
> CLIENT error: rc: -1 errno:115 tcp_rc: 115
> CLIENT error: connect success
> 
> CLIENT error: rc: 1
> CLIENT1 connected
> ========== mg_connect start
> CLIENT has been initialized
> CLIENT error: rc: -1 errno:115 tcp_rc: 115
> CLIENT error: connect success
> 
> CLIENT error: rc: 1
> CLIENT1 connected

Why does the loop start at line 10 ?

The second question：

1. else if (ev == MG_EV_CONNECT) 
2. {
3. #if MG_ENABLE_MBEDTLS || MG_ENABLE_OPENSSL
4. struct mg_tls_opts opts = {.ca = "ss_ca.pem"};
5. mg_tls_init(c, &opts);
6. MG_INFO(("CLIENT initialized TLS"));
7. #endif
8.     *i = 50;
9.     strcpy(tcp_data,"CLIENT connected\r\n");
10.     ql_uart_write(QL_MAIN_UART_PORT, tcp_data, strlen(tcp_data));
11.     //bzero(tcp_data,sizeof(tcp_data));
12.     mg_send(c, "Hi, there", 9);
13. 
14. }

Why does mg_send fail ？

**Comment by scaprile** [comment]
I'm having a terrible time trying to see your code and guess your questions.
What is a non-mainstream RTOS mg_connect ? Are you trying to run Mongoose on some RTOS we don't know ? Did you read the docs ? 
What is your adaptation layer ? Which RTOS ? What exactly are you asking ?

Please follow the guidelines in the [documentation](https://mongoose.ws/documentation)
If there is something that you can't do, please describe it with the smallest piece of code possible.

wrt your second question, or what I think it is:
Many [examples](https://github.com/cesanta/mongoose/tree/master/examples) and [tutorials](https://mongoose.ws/tutorials/) are available.

My bet is that you are expecting that your call to `mg_send()` actually did something, it won't.
See how the examples handle TLS startup, read the tutorials, data is actually sent when the event manager runs, TLS hasn't completed its handshake when you call `mg_send()`
https://mongoose.ws/documentation/tutorials/tls/#tls-for-clients

**Reply by evercf** [reply]
The first problem is that when using **mg_connect()** it keeps repeating from the first line of code, and I do this when I don't use **mg_mgr_poll()**,Repeating when calling the event manager is a procedure such as MG_EV_OPEN->MG_EV_CONNECT. MG_EV_POLL is never triggered when mg_mgr_poll() is used.
Second question, mg_send is run under MG_EV_CONNECT, so tls is not used by default. I succeeded in Linux

**Reply by evercf** [reply]
mg_connect_resolved function in mg_connect returned -1 and error code 115 when connect was connected. I simply added select in mg_connect_resolved and resolved it. Will this change affect mg_send? Without select you will always Need to wait for TCP handshake

**Comment by scaprile** [comment, marked as answer]
Let me put it this way:

>You are not just _using_ Mongoose, you are _porting_ it to a different environment. You don't tell what it is. We don't know and we didn't write the code for it. You are on your own.

> Go to the `src` directory, look at the `arch_*` files, those are support for different environments, see what you can do for your case.

> Look at the `net_*` files, those are support for different TCP/IP stacks, see what you can do for your case.

> Look at `sock.c` there are some special cases handled there.

If you have a specific issue, open an issue, by all means fill the issue template and provide all necessary information. We can't fix your code nor write a port for your RTOS, particularly if we don't know which one it is. Ditto for your TCP/IP stack.

If you are a paying customer or want commercial support, please write to support@cesanta.com

**Reply by evercf** [reply]
I'm using threadx RTOS, I didn't set MG_ENABLE_TCPIP to 1

**Reply by scaprile** [reply]
ThreadX should work with AzureRTOS architecture `#define MG_ARCH MG_ARCH_AZURERTOS`.
Make sure your device is not resetting (code does not magically re-run itself, functions don't jump back to start)

> If you have a specific issue, open an issue, by all means fill the issue template and provide all necessary information.
> If you are a paying customer or want commercial support, please write to support@cesanta.com

**Reply by evercf** [reply]
bool mg_send(struct mg_connection *c, const void *buf, size_t len) {
  quec_uart_test_tcp();
  if (c->is_udp) {
    long n = mg_io_send(c, buf, len);
    MG_DEBUG(("%lu %p %d:%d %ld err %d", c->id, c->fd, (int) c->send.len,
              (int) c->recv.len, n, MG_SOCK_ERR(n)));
    iolog(c, (char *) buf, n, false);
    return n > 0;
  } else {
    return mg_iobuf_add(&c->send, c->send.len, buf, len);
  }
}

c->is_udp=0 in the above source code, so the else function is taken, so the send fails? if I set if(1), it's going to work, so if I'm TCP, is it not going to use this function? I see mg_send has two.

**Comment by evercf** [comment]
Do you see my question? Thank you
